<h3 align="center">Codebasics ML Course: XGBoost Regressor: California Housing Prediction</h3>

### Load Data

In [3]:
import pandas as pd
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()

In [2]:
dir(housing)

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']

In [25]:
print(housing.DESCR)

.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

    :Number of Instances: 20640

    :Number of Attributes: 8 numeric, predictive attributes and the target

    :Attribute Information:
        - MedInc        median income in block group
        - HouseAge      median house age in block group
        - AveRooms      average number of rooms per household
        - AveBedrms     average number of bedrooms per household
        - Population    block group population
        - AveOccup      average number of household members
        - Latitude      block group latitude
        - Longitude     block group longitude

    :Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived

In [4]:
df = pd.DataFrame(housing.data,columns=housing.feature_names)
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [5]:
housing.target_names

['MedHouseVal']

In [6]:
df['MedHouseVal'] = housing.target
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedHouseVal  
0    -122.23        4.526  
1    -122.22        3.585  
2    -122.24        3.521  
3    -122.25        3.413  
4    -122.25        3.422

In [9]:
df.shape

(20640, 9)

In [13]:
df.describe()

MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude   MedHouseVal  
count  20640.000000  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704      2.068558  
std       10.386050      2.135952      2.003532      1.153956  
min        0.692308     32.540000   -124.350000      0.149990  
25%        2.429741     33.930000   -121.800000      1.196000  
50%        2.818116     34.260000   -118.490000      1.797000  
75%        3.282261     37.710000   -118.010000      2.647250  
max     1243.333333     41.950000   -114.310000      5.000010

In [7]:
df.isnull().sum()

MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64

### Train Test Split

In [60]:
from sklearn.model_selection import train_test_split

X = df.drop('MedHouseVal', axis="columns")
y = df['MedHouseVal']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

### Train Linear Regression Model

In [61]:
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
import time

start = time.time()
model = LinearRegression()
model.fit(X_train, y_train)
end= time.time()

y_pred = model.predict(X_test)
r2s = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)

print("R2 Score:", r2s, "MSE:",mse, "time: ",end-start)

R2 Score: 0.596596837481235 MSE: 0.5291402345397312 time:  0.0


### Train Gradient Boosting Regressor

In [62]:
from sklearn.ensemble import GradientBoostingRegressor

start = time.time()
model = GradientBoostingRegressor(n_estimators=100)
model.fit(X_train, y_train)
end = time.time()

y_pred = model.predict(X_test)
r2s = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)

print("R2 Score:", r2s, "MSE:",mse, "time: ",end-start)

R2 Score: 0.7790269318453287 MSE: 0.28984835017223687 time:  6.145176887512207


### Train XGBoost Regressor

In [63]:
from xgboost import XGBRegressor

# model = XGBRegressor(objective ='reg:squarederror', colsample_bytree = 0.3, learning_rate = 0.1,
#                          max_depth = 5, alpha = 10, n_estimators = 100)
start = time.time()
model = XGBRegressor(n_estimators=100)
model.fit(X_train, y_train)
end = time.time()

y_pred = model.predict(X_test)
r2s = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)

print("R2 Score:", r2s, "MSE:",mse, "time: ",end-start)

R2 Score: 0.8326512949840744 MSE: 0.21950976405131195 time:  0.12275862693786621


You can see that from Linear regression performs the worst (R2 score: 0.59), whereas gradient boosting model shows significant improvement (R2 score: 0.77). When we use XGBoost the performance improves further (R2 score: 0.83)

Time wise XGBoost runs way faster (0.12 sec) compared to gradient boosting (6 seconds)